go/types/opcache.go

```diff
@@ -80,144 +147,185 @@ func (store *ldbOpCacheStore) opCache() opCache {
+		var gk Value
+		ldbVal, gk = decodeValue(ldbVal, true, i.vr)
+		encodedValues = append(encodedValues, gk)
+	}
 
-// Insert can be called from any goroutine
-func (opc *ldbOpCache) SetInsert(val Value) {
-	switch val.Type().Kind() {
-	default:
-		ldbKey := ldbKeyFromValueHash(val, opc.colId)
-		valArray := [initialBufferSize]byte{}
-		encodedVal := encToSlice(val, valArray[:], opc.vrw)
-		err := opc.ldb.Put(ldbKey, encodedVal, nil)
-		d.Chk.NoError(err)
+	// Fold in any non-primitive key values that were stored in ldbVal
+	keys := ValueSlice{}
+	pos := 0
+	for _, k1 := range encodedKeys {
+		if k1 == nil {
+			k1 = encodedValues[pos]
+			pos++
+		}
+		keys = append(keys, k1)
+	}
 
-	case BoolKind, NumberKind, StringKind:
-		ldbKey := ldbKeyFromValue(val, opc.colId, opc.vrw)
-		// Since the ldbKey contains the val, there's no reason to store anything in the ldbValue
-		err := opc.ldb.Put(ldbKey, nil, nil)
-		d.Chk.NoError(err)
+	key := keys[len(keys)-1]
+	graphKeys := keys[:len(keys)-1]
+	var item sequenceItem
+	switch opKind {
+	case MapKind:
+		val := encodedValues[len(encodedValues)-1]
+		item = mapEntry{key, val}
+	case SetKind:
+		item = key
+	case ListKind:
+		item = encodedValues[len(encodedValues)-1]
 	}
+
+	return graphKeys, opKind, item
 }
 
-func (i *ldbOpCacheIterator) SetOp() sequenceItem {
-	ldbKey := i.iter.Key()
+func (opc *ldbOpCache) NewIterator() opCacheIterator {
+	prefix := [4]byte{}
+	binary.BigEndian.PutUint32(prefix[:], opc.colId)
+	return &ldbOpCacheIterator{iter: opc.ldb.NewIterator(util.BytesPrefix(prefix[:]), nil), vr: opc.vrw}
+}
 
-	switch NomsKind(ldbKey[uint32Size]) {
-	case BoolKind, NumberKind, StringKind:
-		return DecodeFromBytes(ldbKey[uint32Size:], i.vr, staticTypeCache)
-	default:
-		data := i.iter.Value()
-		return DecodeFromBytes(data, i.vr, staticTypeCache)
-	}
+func (i *ldbOpCacheIterator) Next() bool {
+	return i.iter.Next()
 }
 
 func (i *ldbOpCacheIterator) Release() {
 	i.iter.Release()
 }
 
-// writeLdbKeyHeaderBytes writes the first 4 or 5 bytes into the ldbKey. The first 4 bytes in every
-// ldbKey are the colId. This identifies all the keys for a particular opCache and allows this opStore
-// to cache values for multiple collections. The optional 5th byte is the NomsKind of the value. In
-// cases where we're encoding the hash of an object we need to write the nomsKind manually because the
-// hash doesn't contain it. In cases were we are encoding a primitive value into the key, the first byte
-// of the value is the nomsKind so there is no reason to write it again.
-func writeLdbKeyHeaderBytes(ldbKey []byte, colId uint32, v Value) []byte {
-	binary.LittleEndian.PutUint32(ldbKey, colId)
-	length := uint32Size
-	if v != nil {
-		ldbKey[length] = byte(v.Type().Kind())
-		length++
+func encodeKeys(bs []byte, opKind NomsKind, keys []Value, vrw ValueReadWriter) ([]byte, []Value) {
+	// All ldb keys start with a 4-byte collection id that serves as a namespace
```

Comment:
Decided to move the insertion of the first colId into the function to keep it all together.